tiledb/sm/object/object_iter.cc

```diff
@@ -0,0 +1,220 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements object iterator functions.
+ */
+
+#include "tiledb/sm/object/object_iter.h"
+#include "tiledb/sm/enums/object_type.h"
+#include "tiledb/sm/object/object.h"
+
+namespace tiledb::sm {
+
+class ObjectIterException : public StatusException {
+ public:
+  explicit ObjectIterException(const std::string& message)
+      : StatusException("ObjectIter", message) {
+  }
+};
+
+ObjectIter* object_iter_begin(
+    ContextResources& resources, const char* path, WalkOrder order) {
+  // Sanity check
+  URI path_uri(path);
+  if (path_uri.is_invalid()) {
+    throw ObjectIterException(
+        "Cannot create object iterator; Invalid input path");
+  }
+
+  // Get all contents of path
+  std::vector<URI> uris;
+  throw_if_not_ok(resources.vfs().ls(path_uri, &uris));
+
+  // Create a new object iterator
+  ObjectIter* obj_iter = tdb_new(ObjectIter);
+  obj_iter->order_ = order;
+  obj_iter->recursive_ = true;
+
+  // Include the uris that are TileDB objects in the iterator state
+  ObjectType obj_type;
+  for (auto& uri : uris) {
+    auto st = object_type(resources, uri, &obj_type);
+    if (!st.ok()) {
+      tdb_delete(obj_iter);
+      std::throw_with_nested(ObjectIterException(
+          "Cannot create object iterator; " + st.message()));
+    }
+    if (obj_type != ObjectType::INVALID) {
+      obj_iter->objs_.push_back(uri);
+      if (order == WalkOrder::POSTORDER)
+        obj_iter->expanded_.push_back(false);
+    }
+  }
+
+  return obj_iter;
+}
+
+ObjectIter* object_iter_begin(ContextResources& resources, const char* path) {
+  // Sanity check
+  URI path_uri(path);
+  if (path_uri.is_invalid()) {
+    throw ObjectIterException(
+        "Cannot create object iterator; Invalid input path");
+  }
+
+  // Get all contents of path
+  std::vector<URI> uris;
+  throw_if_not_ok(resources.vfs().ls(path_uri, &uris));
+
+  // Create a new object iterator
+  ObjectIter* obj_iter = tdb_new(ObjectIter);
+  obj_iter->order_ = WalkOrder::PREORDER;
+  obj_iter->recursive_ = false;
+
+  // Include the uris that are TileDB objects in the iterator state
+  ObjectType obj_type;
+  for (auto& uri : uris) {
+    throw_if_not_ok(object_type(resources, uri, &obj_type));
```

Comment:
This is a preexisting defect. `obj_iter` will leak if this statement throws.

It looks like it's probably best to keep `throw_if_not_ok` and to put it into a `try` block that calls `tdb_delete` in the `catch`. This would be a situation where `throw_with_nested` would be used, but I'm not convinced that we need it and couldn't just use unadorned `throw` instead (particularly to keep the changes to a minimum in this PR).

Whatever you do here, do the same thing in the corresponding code in the variation immediately above. We need to remove this copypasta, but this isn't the right PR to do it.